src/core/execute.c

```diff
@@ -1213,6 +1213,10 @@ static int setup_pam(
 
                 if (getttyname_malloc(STDIN_FILENO, &q) >= 0)
                         tty = strjoina("/dev/", q);
+                else
+                        /* If everything else failed then let's just use value "systemd". This will cause that session
+                         * isn't going to be marked as "background" and user manager will be started. */
+                        tty = "systemd";
```

Comment:
This sounds hackish for me...
At least, `pam_sm_open_session()` in pam_systemd.c needs to handle this gracefully.

Reply:
We have conflict of requirements. I don't want background sessions (of cron) to spawn user instance while at the same time are relying on (currently) background sessions created from sd_bus to spawn user manager so that connection to user bus can be established.

https://github.com/systemd/systemd/blob/3bdb60055e16776c8cc4d44f26abb5f2cbe197bf/src/libsystemd/sd-bus/sd-bus.c#L1567

I am essentially saying here that no PAM sessions created from systemd through `PAMName=` directive will be considered background (this was effectively the previous behaviour even though they were marked as background).

Reply:
yeah, I understand the motivation of this change. How about the following?
```diff
diff --git a/src/login/pam_systemd.c b/src/login/pam_systemd.c
index 2047dc1817..1519b25182 100644
--- a/src/login/pam_systemd.c
+++ b/src/login/pam_systemd.c
@@ -792,6 +792,10 @@ _public_ PAM_EXTERN int pam_sm_open_session(
                              * much later (this is because it doesn't know yet if it needs one at all, as whether to
                              * register a pty or not is negotiated much later in the protocol). */
 
+        } else if (streq(tty, "systemd")) {
+                if (isempty(class))
+                        class = "user";
+                tty = NULL;
         } else
                 /* Chop off leading /dev prefix that some clients specify, but others do not. */
                 tty = skip_dev_prefix(tty);
```

Reply:
This seems sensible. Thank you for the suggestion.